core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

```diff
@@ -377,33 +376,39 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
     }
   }
 
-  private def validateNoWorkspaces(projectName: RawlsBillingProjectName): Future[Unit] = {
-    for {
-      workspaceCount <- dataSource.inTransaction { dataAccess =>
-        dataAccess.workspaceQuery.countByNamespace(projectName)
-      }
-      _ <- if (workspaceCount > 0) {
-        Future.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.BadRequest, "Project cannot be deleted because it contains workspaces.")))
-      } else {
-        Future.successful(())
-      }
-    } yield ()
-  }
+  private def failUnlessHasNoWorkspaces(projectName: RawlsBillingProjectName): Future[Unit] =
+    dataSource.inTransaction(_.workspaceQuery.countByNamespace(projectName)) map { count =>
+      if (count == 0) () else throw new RawlsExceptionWithErrorReport(ErrorReport(
+        StatusCodes.BadRequest,
+        "Project cannot be deleted because it contains workspaces."
+      ))
+    }
 
   // TODO - once workspace migration is complete and there are no more v1 workspaces or v1 billing projects, we can remove this https://broadworkbench.atlassian.net/browse/CA-1118
   private def deleteGoogleProjectIfChild(projectName: RawlsBillingProjectName, userInfoForSam: UserInfo, deleteGoogleProjectWithGoogle: Boolean = true) = {
-    samDAO.listResourceChildren(SamResourceTypeNames.billingProject, projectName.value, userInfoForSam).flatMap { projectChildren =>
-      if (projectChildren.contains(SamFullyQualifiedResourceId(projectName.value, SamResourceTypeNames.googleProject.value))) {
+    def googleProjectExists(projectId: GoogleProjectId) =
+      gcsDAO.getGoogleProject(projectId) transform {
+        case Success(_) => Success(true)
+        case Failure(e: HttpResponseException) if e.getStatusCode == 404 => Success(false)
+        case Failure(t) => Failure(t)
+      }
+
+    def F = Applicative[Future]
```

Comment:
I have some minor reservations about introducing Applicative here. I'm not sure what benefits it provides over an if statement. Is there something I'm missing?